core/store/src/db.rs

```diff
@@ -830,7 +832,22 @@ fn rocksdb_column_options(col: DBCol) -> Options {
     opts
 }
 
+// Number of RocksDB instances in the process.
+pub static ROCKSDB_INSTANCES_COUNTER: Lazy<(Mutex<usize>, Condvar)> =
+    Lazy::new(|| (Mutex::new(0), Condvar::new()));
+
 impl RocksDB {
+    /// Blocks until all RocksDB instances (usually 0 or 1) gracefully shutdown.
+    pub fn block_until_all_instances_are_dropped() {
+        debug!(target: "db", "Blocking until all RocksDB instances perform graceful shutdown");
+        let (lock, cvar) = &*ROCKSDB_INSTANCES_COUNTER;
+        let mut num_instances = lock.lock().unwrap();
+        while *num_instances != 0 {
+            num_instances = cvar.wait(num_instances).unwrap();
```

Comment:
Might want to add a log line here -- it feels like the place where we can get stuck if, for some reason, some of the threads *don't* shut down.